src/plugins/data_source/server/saved_objects/credential_saved_objects_client_wrapper.ts

```diff
@@ -0,0 +1,326 @@
+
+import { Credential } from '../../common';
+
+/**
+ * Describes the Credential Saved Objects Client Wrapper class,
+ * which contains the factory used to create Saved Objects Client Wrapper instances
+ */
+export class CredentialSavedObjectsClientWrapper {
+  private type: string = 'credential';
+  private authType: Credential.SharedAuthType = 'shared';
+  private noAuthType: Credential.NoAuthType = 'no_auth';
+  private usernamePasswordType: Credential.UsernamePasswordType = 'username_password';
+
+  constructor(private cryptographyClient: CryptographyClient) {}
+
+  /**
+   * Describes the factory used to create instances of Saved Objects Client Wrappers
+   * for credential spcific operations such as encryption
+   * Check {@link Credential.CredentialSavedObjectAttributes} for attributes type details
+   * Check {@link Credential.CredentialMaterials} for credential materials type and
+   * credential materials content details
+   */
+  public wrapperFactory: SavedObjectsClientWrapperFactory = (wrapperOptions) => {
+    const createWithCredentialMaterialsEncryption = async <T = unknown>(
+      type: string,
+      attributes: T,
+      options?: SavedObjectsCreateOptions
+    ) => {
+      if (this.type !== type) {
+        return await wrapperOptions.client.create(type, attributes, options);
+      }
+
+      this.validateAttributes(attributes);
+
+      const { credentialMaterials } = attributes;
+
+      if (this.noAuthType === credentialMaterials.credentialMaterialsType) {
+        const noAuthAttributes = {
+          ...attributes,
+          credentialMaterials: this.dropCredentialMaterialsContent(credentialMaterials),
+        };
+        return await wrapperOptions.client.create(type, noAuthAttributes, options);
+      }
+
+      if (this.usernamePasswordType === credentialMaterials.credentialMaterialsType) {
+        const encryptedAttributes = {
+          ...attributes,
+          credentialMaterials: await this.validateAndEncryptCredentialMaterials(
+            credentialMaterials
+          ),
+        };
+        return await wrapperOptions.client.create(type, encryptedAttributes, options);
+      }
+
+      throw SavedObjectsErrorHelpers.createBadRequestError(
+        `Invalid credential materials type: '${credentialMaterials.credentialMaterialsType}'`
+      );
+    };
+
+    const bulkCreateWithCredentialMaterialsEncryption = async <T = unknown>(
+      objects: Array<SavedObjectsBulkCreateObject<T>>,
+      options?: SavedObjectsCreateOptions
+    ): Promise<SavedObjectsBulkResponse<T>> => {
+      objects = await Promise.all(
+        objects.map(async (object) => {
+          const { type, attributes } = object;
+
+          if (this.type !== type) {
+            return object;
+          }
+
+          this.validateAttributes(attributes);
+
+          const { credentialMaterials } = attributes;
+
+          if (this.noAuthType === credentialMaterials.credentialMaterialsType) {
```

Comment:
we need to think a bit more about no auth type, e.g. in case of user choose to use no auth type for a data source, shall we link the data source with a credential which doesn't have any actual credential? or we do not need to let data source to refer to a credential. If we go with the second route, we may have an auth type, or auth/noauth flag in data source

Reply:
Yes, we started this conversation with @KrooshalUX. From an engineering standpoint, adding authType to the data source makes more sense for single responsibility. Will update the logic for credential side.